### PR TITLE
RSM bytestream compatible to 2.3

### PR DESCRIPTION
**Pull request by kokosha**
Update the RSM bytestream to be compatible with version 2.2 and 2.3.

### CODE REVIEW

File: korangar/src/loaders/model/mod.rs [hunk 1]
```diff
@@ -116,10 +116,10 @@ impl ModelLoader {
 
     fn calculate_matrices(node: &NodeData, parent_matrix: &Matrix4<f32>) -> (Matrix4<f32>, Matrix4<f32>, Matrix4<f32>) {
         let main = Matrix4::from_translation(node.translation1) * Matrix4::from(node.offset_matrix);
-
-        let scale_matrix = Matrix4::from_nonuniform_scale(node.scale.x, node.scale.y, node.scale.z);
-        let rotation_matrix = Matrix4::from_axis_angle(node.rotation_axis, Rad(node.rotation_angle));
-        let translation_matrix = Matrix4::from_translation(node.translation2);
+        let scale_size = node.scale.unwrap();
```
Comment (vE5li) [inline korangar/src/loaders/model/mod.rs:119]: `scale_size` doesn't really make a lot of sense, just `scale` would be better

File: ragnarok_formats/src/model.rs [hunk 2]
```diff
@@ -57,21 +57,64 @@ where
...
+#[derive(Debug, ByteConvertable)]
+#[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
+pub struct TextureKeyframeData {
+    // List of operation types - https://rathena.org/board/topic/127587-rsm2-file-format/
+    // 0: Texture translation on the X axis. The texture is tiled.
+    // 1: Texture translation on the Y axis. The texture is tiled.
+    // 2: Texture multiplication on the X axis. The texture is tiled.
+    // 3: Texture multiplication on the Y axis. The texture is tiled
+    // 4: Texture rotation around (0, 0). The texture is not tiled.
+    pub operation_type: u32,
```
Comment (vE5li) [inline ragnarok_formats/src/model.rs:99]: Could we make this be an enum?

File: ragnarok_formats/src/model.rs [hunk 3]
```diff
@@ -115,15 +168,28 @@ pub struct NodeData {
...
-    #[repeating_option(position_keyframe_count)]
-    pub position_keyframes: Vec<PositionKeyframeData>,
+    pub scale_keyframe_count: Option<u32>,
+    #[repeating_option(scale_keyframe_count)]
+    pub scale_keyframes: Vec<ScaleKeyframeData>,
     #[new_derive]
     pub rotation_keyframe_count: u32,
     #[repeating(rotation_keyframe_count)]
     pub rotation_keyframes: Vec<RotationKeyframeData>,
+    // TODO: Translation Keyframe not implemented yet
```
Comment (vE5li) [inline ragnarok_formats/src/model.rs:181]: What do you mean by not yet implemented? The parsing or using the data?

**Review by vE5li — COMMENTED** [review]
Looks good! Just some small comments/questions

File: ragnarok_formats/src/model.rs (hunk 3 above)
Comment (kokosha) [reply ragnarok_formats/src/model.rs:181]: The manipulation of the model

File: ragnarok_formats/src/model.rs (hunk 2 above)
Comment (kokosha) [reply ragnarok_formats/src/model.rs:99]: I will change to use a enum

File: ragnarok_formats/src/model.rs (hunk 3 above)
Comment (vE5li) [reply ragnarok_formats/src/model.rs:181]: The we should either remove the comments or put it somewhere else. Something like `TODO: Use translations keyframes` in the `load` method of the `ModelLoader` is less likely to become a stale comment.

File: ragnarok_formats/src/model.rs [hunk 4]
```diff
@@ -57,21 +57,74 @@ where
...
 pub struct RotationKeyframeData {
     pub frame: u32,
     pub quaternions: Quaternion<f32>,
 }
 
+#[derive(Debug, ByteConvertable)]
+#[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
+pub struct TranslationKeyframeData {
+    pub frame: u32,
+    pub translate: Vector3<f32>,
```
Comment (vE5li) [inline ragnarok_formats/src/model.rs:76]: Given that the struct is called `TranslationKeyframeData`, I think this should be `translation`

File: korangar/src/loaders/model/mod.rs [hunk 5]
```diff
@@ -247,6 +247,9 @@ impl ModelLoader {
         };
 
         // TODO: Temporary check until we support more versions.
+        // TODO: The model operation to scale keyframe not implemented yet
+        // TODO: The model operation to translate keyframe not implemented yet
+        // TODO: The model operation to modify texture keyframe not implemented yet
```
Comment (vE5li) [inline korangar/src/loaders/model/mod.rs:252]: Please make sure that the comments use correct grammar (and end them on a `.` please)

File: ragnarok_formats/src/model.rs [hunk 6]
```diff
@@ -57,21 +57,74 @@ where
...
+// 4: Texture rotation around (0, 0). The texture is not tiled.
+#[derive(Debug, ByteConvertable)]
+#[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
+pub enum TextureOperation {
+    TranslationX,
+    TranslationY,
+    ScaleX,
+    ScaleY,
+    Rotation,
+}
```
Comment (vE5li) [inline ragnarok_formats/src/model.rs:104]: For documentation style comments like this you should really use Rusts doc comments (`///`). That will make these comments show up in peoples IDEs as well as documentation generated by `cargo doc`.

You can attach doc comments to almost anything, for example the varians of an enum. So I would instead write the coments like this:

```rust
/// List of texture operation types.
/// See: https://rathena.org/board/topic/127587-rsm2-file-format/
#[derive(Debug, ByteConvertable)]
#[cfg_attr(feature = "interface", derive(korangar_interface::elements::PrototypeElement))]
pub enum TextureOperation {
    /// Texture translation on the X axis. The texture is tiled.
    TranslationX,
    /// Texture translation on the Y axis. The texture is tiled.
    TranslationY,
    /// Texture multiplication on the X axis. The texture is tiled.
    ScaleX,
    /// Texture multiplication on the Y axis. The texture is tiled
    ScaleY,
    /// Texture rotation around (0, 0). The texture is not tiled.
    Rotation,
}
```

File: ragnarok_formats/src/model.rs [hunk 7]
```diff
@@ -57,27 +57,84 @@ where
...
+    #[version_equals_or_above(2, 2)]
+    pub length: Option<u32>,
     pub vertex_position_indices: [u16; 3],
     pub texture_coordinate_indices: [u16; 3],
     pub texture_index: u16,
     pub padding: u16,
     pub two_sided: i32,
     pub smooth_group: i32,
+    #[version_equals_or_above(2, 2)]
+    #[repeating_expr(usize::max(0, (length.unwrap() as usize - 24)/4))]
```
Comment (vE5li) [inline ragnarok_formats/src/model.rs:136]:
```suggestion
    #[repeating_expr(0.max((length.unwrap() as usize - 24) / 4))]
```

File: ragnarok_formats/src/model.rs [hunk 8]
```diff
@@ -92,17 +149,32 @@ pub struct TextureCoordinateData {
...
-    pub scale: Vector3<f32>,
+    #[version_smaller(2, 2)]
+    #[new_default]
+    pub rotation_angle: Option<f32>,
+    #[version_smaller(2, 2)]
+    #[new_default]
+    pub rotation_axis: Option<Vector3<f32>>,
+    #[version_smaller(2, 2)]
+    #[new_default]
+    pub scale: Option<Vector3<f32>>,
```
Comment (vE5li) [inline ragnarok_formats/src/model.rs:177]: Why are `translation1`, `rotation_angle`, `rotation_axis` and `scale` marked with `new_default`?

`new_default` is meant for fields that the user should never care about setting when creating a new instance of the type (like the signature, unknown bytes, patting, etc.).

EDIT: Actually, same for `frames_per_second`

**Comment by vE5li** [comment]
After you applied those changes please squash the commit history so I can merge :)

File: ragnarok_formats/src/model.rs [hunk 9]
```diff
@@ -57,27 +57,84 @@ where
...
+    #[version_equals_or_above(2, 2)]
+    pub length: Option<u32>,
     pub vertex_position_indices: [u16; 3],
     pub texture_coordinate_indices: [u16; 3],
     pub texture_index: u16,
     pub padding: u16,
     pub two_sided: i32,
     pub smooth_group: i32,
+    #[version_equals_or_above(2, 2)]
+    #[repeating_expr(0.max((length.unwrap() as usize - 24) / 4))]
```
Comment (vE5li) [inline ragnarok_formats/src/model.rs:136]: My bad, I just realized the `0.max` is also redundant. The type is already `usize` so it can't be negative

**Pull request merged**